driver-core/src/main/com/mongodb/internal/operation/ClientBulkWriteOperation.java

```diff
@@ -180,25 +191,65 @@ public ClientBulkWriteResult execute(final WriteBinding binding) throws ClientBu
+        WriteConcern effectiveWriteConcern = validateAndGetEffectiveWriteConcern(binding.getOperationContext().getSessionContext());
+        ResultAccumulator resultAccumulator = new ResultAccumulator();
+        MutableValue<MongoException> transformedTopLevelError = new MutableValue<>();
+
+        beginAsync().<Void>thenSupply(c -> {
+            executeAllBatchesAsync(effectiveWriteConcern, binding, resultAccumulator, c);
+        }).onErrorIf(topLevelError -> topLevelError instanceof MongoException, (topLevelError, c) -> {
+            transformedTopLevelError.set(transformWriteException((MongoException) topLevelError));
+            c.complete(c);
+        }).<ClientBulkWriteResult>thenApply((ignored, c) -> {
+            c.complete(resultAccumulator.build(transformedTopLevelError.getNullable(), effectiveWriteConcern));
+        }).finish(finalCallback);
+    }
+
     /**
      * To execute a batch means:
      * <ul>
      *     <li>execute a `bulkWrite` command, which creates a cursor;</li>
      *     <li>consume the cursor, which may involve executing `getMore` commands.</li>
      * </ul>
      *
-     * @throws MongoException When a {@linkplain ClientBulkWriteException#getError() top-level error} happens.
+     * @throws MongoException When a {@linkplain ClientBulkWriteException#getCause() top-level error} happens.
      */
     private void executeAllBatches(
             final WriteConcern effectiveWriteConcern,
             final WriteBinding binding,
             final ResultAccumulator resultAccumulator) throws MongoException {
-        Integer nextBatchStartModelIndex = 0;
+        Integer nextBatchStartModelIndex = INITIAL_BATCH_MODEL_START_INDEX;
+
         do {
             nextBatchStartModelIndex = executeBatch(nextBatchStartModelIndex, effectiveWriteConcern, binding, resultAccumulator);
         } while (nextBatchStartModelIndex != null);
     }
 
+    /**
+     * @see #executeAllBatches(WriteConcern, WriteBinding, ResultAccumulator)
+     */
+    private void executeAllBatchesAsync(
+            final WriteConcern effectiveWriteConcern,
+            final AsyncWriteBinding binding,
+            final ResultAccumulator resultAccumulator,
+            final SingleResultCallback<Void> finalCallback) {
+        MutableValue<Integer> nextBatchStartModelIndex = new MutableValue<>(INITIAL_BATCH_MODEL_START_INDEX);
+
+        beginAsync().thenRunDoWhileLoop(iterationCallback -> {
+            beginAsync().<Integer>thenSupply(c -> {
+                executeBatchAsync(nextBatchStartModelIndex.get(), effectiveWriteConcern, binding, resultAccumulator, c);
+            }).<Void>thenApply((nextBatchStartModelIdx, c) -> {
+                nextBatchStartModelIndex.set(nextBatchStartModelIdx);
+                c.complete(c);
+            }).finish(iterationCallback);
+        }, () -> nextBatchStartModelIndex.getNullable() != null)
+          .finish(finalCallback);
```

Comment:
Let's re-format in accordance with the `beginAsync` docs:

```suggestion
        }, () -> nextBatchStartModelIndex.getNullable() != null
        ).finish(finalCallback);
```